docs/proposals/20210301-enhancement_of_YurtHub_caching_ability.md

```diff
@@ -0,0 +1,257 @@
+
+- `CanCacheFor(req *http.Request)` function uses a resourceToKindMap to determine if the response message needs to be cached;
+
+- `SaveOneObject(...)`, `SaveListObject(...)` and `SaveWatchObject(...)` use resourceToKindMap to obtain the kind information corresponding to the resource name, which is used to cache the response message;
+
+  Because the resolved object does not contain GVK information. But when caching objects, GVK information needs to be set. The above two maps pre-store some resource-to-kind mappings, which can be used to cache resources and retrieve the corresponding kind based on the resource name in the request.URL .
+
+- The `queryListObject(...)` function uses the resourceToListKindMap member to query the List object
+
+Due to the above limitations, Yurthub can only cache resources defined in [resourceToKindMap](https://github.com/openyurtio/openyurt/blob/4d7463a40801c29d09c4f7d10ba46b73cb019915/pkg/yurthub/cachemanager/cache_manager.go#L46) and  [resourceToListKindMap](https://github.com/openyurtio/openyurt/blob/4d7463a40801c29d09c4f7d10ba46b73cb019915/pkg/yurthub/cachemanager/cache_manager.go#L62).
+
+## Proposal
+
+### Delete [resourceToKindMap](https://github.com/openyurtio/openyurt/blob/4d7463a40801c29d09c4f7d10ba46b73cb019915/pkg/yurthub/cachemanager/cache_manager.go#L46) and [resourceToListKindMap](https://github.com/openyurtio/openyurt/blob/4d7463a40801c29d09c4f7d10ba46b73cb019915/pkg/yurthub/cachemanager/cache_manager.go#L62)
+
+Remove the restriction that only certain resources can be cached by removing [resourceToKindMap](https://github.com/openyurtio/openyurt/blob/4d7463a40801c29d09c4f7d10ba46b73cb019915/pkg/yurthub/cachemanager/cache_manager.go#L46) and [resourceToListKindMap](https://github.com/openyurtio/openyurt/blob/4d7463a40801c29d09c4f7d10ba46b73cb019915/pkg/yurthub/cachemanager/cache_manager.go#L62). Removing the above restriction requires solving the problem which is preserving the original GVK information while decoding the corresponding resources;
+
+The reason why object misses the field of apiverison and kind is YurtHub using [serializer.WithoutConversionCodecFactory](https://github.com/alibaba/openyurt/blob/7095962687e2666a27d14e436b96ef893b228beb/pkg/yurthub/kubernetes/serializer/serializer.go#L48) as NegotiatedSerializer, which produces [WithoutVersionDecoder ](https://github.com/kubernetes/kubernetes/blob/994b5c6cc20bda188ac9bc33217f2dbc7ecf45bb/staging/src/k8s.io/apimachinery/pkg/runtime/helper.go#L256)and remove the apiversion/kind of object when serialized from http.response. The following code is the decoder that is invoked when the resource is decoded:
+
```

Comment:
What's the meaning of removing the apiversion/kind of object when use `WithoutVersionDecoder`?

Reply:
Currently, YurtHub use WithoutConversionCodecFactory as NegotiatedSerializer. Because the WithoutConversionCodecFactory  is a CodecFactory that will explicitly ignore requests to perform conversion, the GVK information will be ignored when this WithoutConversionCodecFactory  use the decoder(which is [Decoder](https://github.com/kubernetes/kubernetes/blob/994b5c6cc20bda188ac9bc33217f2dbc7ecf45bb/staging/src/k8s.io/apimachinery/pkg/runtime/helper.go#L256)
So we're currently using the new CodecFactory wrapper which is [WithVersionCodecFactory](https://github.com/qclc/openyurt/blob/9aaa2386dd68d2dbc703e56dbc88852fe9e116b1/pkg/yurthub/kubernetes/serializer/serializer.go#L41) as negotiatedSerializer, and rewriting the original [Decode](https://github.com/qclc/openyurt/blob/9aaa2386dd68d2dbc703e56dbc88852fe9e116b1/pkg/yurthub/kubernetes/serializer/serializer.go#L69) function to keep the GVK information.

Reply:
> Currently, YurtHub use [WithoutConversionCodecFactory](https://github.com/alibaba/openyurt/blob/7095962687e2666a27d14e436b96ef893b228beb/pkg/yurthub/kubernetes/serializer/serializer.go#L48) as NegotiatedSerializer. Because the WithoutConversionCodecFactory is a CodecFactory that will explicitly ignore requests to perform conversion, the GVK information will be ignored when this WithoutConversionCodecFactory use the decoder (which is this [Decoder](https://github.com/kubernetes/kubernetes/blob/994b5c6cc20bda188ac9bc33217f2dbc7ecf45bb/staging/src/k8s.io/apimachinery/pkg/runtime/helper.go#L256)).
> So we're currently adding the new CodecFactory wrapper which is [WithVersionCodecFactory](https://github.com/qclc/openyurt/blob/9aaa2386dd68d2dbc703e56dbc88852fe9e116b1/pkg/yurthub/kubernetes/serializer/serializer.go#L41) to replace the old WithoutConversionCodecFactory as negotiatedSerializer, and rewriting the original [Decode](https://github.com/qclc/openyurt/blob/9aaa2386dd68d2dbc703e56dbc88852fe9e116b1/pkg/yurthub/kubernetes/serializer/serializer.go#L69) function to keep the GVK information.

And when we can use `WithoutVersionDecoder `?